addon/-private/system/relationships/relationships-payloads.js

```diff
@@ -0,0 +1,446 @@
+      let id = work[i][1];
+      let relationshipName = work[i][2];
+      let relationshipData = work[i][3];
+
+      let entry = {
+        payload: relationshipData,
+        inverse: false
+      };
+      // TODO: maybe delay this slightly?
+      let inverseEntry = {
+        payload: {
+          data: {
+            id: id,
+            type: modelName
+          }
+        },
+        inverse: true
+      }
+
+      if (modelName === this._lhsModelName && relationshipName === this._lhsRelationshipName) {
+        this._removeInverse(id, this._lhsPayloads[id], this._rhsPayloads);
+        this._lhsPayloads[id] = entry;
+        this._populateInverse(relationshipData, inverseEntry, this._rhsPayloads, this._rhsRelationshipIsMany);
+      } else {
+        this._removeInverse(id, this._rhsPayloads[id], this._lhsPayloads);
+        this._rhsPayloads[id] = entry;
+        this._populateInverse(relationshipData, inverseEntry, this._lhsPayloads, this._lhsRelationshipIsMany);
+      }
+    }
+  }
+
+  _inverseLoaded(entry) {
+    let data = entry && entry.payload && entry.payload.data;
+    if (!data) { return false; }
+
+    if (Array.isArray(data)) {
+      for (let i=0; i<data.length; ++i) {
+        if (this._hasRecordForId(this._store, data[i].type, data[i].id)) {
+          return true;
+        }
+      }
+      return false;
+    } else {
+      return this._hasRecordForId(this._store, data.type, data.id);
+    }
+  }
+
+  _hasRecordForId(store, type, id) {
+    return typeof type === 'string' &&
+      typeof id !== 'undefined' &&
+      id !== null &&
+      store.hasRecordForId(type, id);
+  }
+
+  _populateInverse(relationshipData, inverseEntry, inversePayloads, inverseIsMany) {
+    if (!relationshipData.data) { return; }
+
+    if (Array.isArray(relationshipData.data)) {
+      for (let i=0; i<relationshipData.data.length; ++i) {
+        let inverseId = relationshipData.data[i].id;
+        this._addToInverse(inverseEntry, inverseId, inversePayloads, inverseIsMany);
+      }
+    } else {
+      let inverseId = relationshipData.data.id;
+      this._addToInverse(inverseEntry, inverseId, inversePayloads, inverseIsMany);
+    }
+  }
+
+  _addToInverse(inverseEntry, inverseId, inversePayloads, inverseIsMany) {
+    let existingEntry = inversePayloads[inverseId];
+    let existingData = existingEntry && existingEntry.payload && existingEntry.payload.data;
+
+    if (Array.isArray(existingData)) {
+      existingData.push(inverseEntry.payload.data);
+    } else {
+      if (inverseIsMany) {
+        inversePayloads[inverseId] = {
+          payload: {
+            data: [inverseEntry.payload.data]
+          },
+          inverse: true
+        }
+      } else {
+        inversePayloads[inverseId] = inverseEntry;
+      }
+    }
+  }
+
+  get _lhsRelationshipIsMany() {
+    if (this.__lhsRelatinoshipIsMany === undefined) {
+      if (this._lhsModelName) {
+        let modelClass = this._store.modelFor(this._lhsModelName);
```

Comment:
> I believe both LHS and RHS modelClass's are loaded when RelationshipPayload is instantiated, should we just pass them in at construction? At that point, do we already have the relationship descriptor as well, or is that what we want lazy?

This is not necessarily the case.  `inverseFor` does not load the inverse class.  Even when we check for the inverse relationship, we won't cause the inverse class to be loaded.